### PR TITLE
Changed the static_init macro to take a size argument

### PR DESCRIPTION
**Pull request by Idolf**
- Changed `static_init!()` to take a size argument. This size argument should ideally be inferred, but for now we write it manually and do a compile-time verification that it is correct.
- Changed the type signature from `pub unsafe fn init<'a>() -> &'a mut Firestorm ` to `pub unsafe fn init() -> &'static mut Firestorm`, since the first is just a more convuluted way of writing the second.

### CODE REVIEW

**Comment by ppannuto** [comment]
Hey, it looks like your editor made a bunch of (good) style changes, but it makes it really hard to see what actually changed.

Would you mind terribly splitting this into two commits, one with just the style stuff and one with the actual changes?

Thanks!

**Comment by Idolf** [comment]
Nope, I don't mind. Doing it now.

File: src/platform/storm/lib.rs [hunk 1]
```diff
@@ -68,17 +69,18 @@ impl Firestorm {
...
 macro_rules! static_init {
-   ($V:ident : $T:ty = $e:expr) => {
-        let $V : &mut $T = {
+   ($V:ident : $T:ty = $e:expr, $size:expr) => {
+        let $V : &'static mut $T = {
             // Waiting out for size_of to be available at compile-time to avoid
             // hardcoding an abitrary large size...
-            static mut BUF : [u8; 1024] = [0; 1024];
+            static mut BUF : [u8; $size] = [0; $size];
+            assert!(mem::size_of::<$T>() <= $size);
```
Comment (alevy) [inline src/platform/storm/lib.rs]: if we're OK with tight bounds, it's actually possible to get the compiler to verify this statically. Something like:

```rust
let _ : T = mem::transmute(BUF);
```

Comment (Idolf) [inline src/platform/storm/lib.rs]: That is a good idea. I will add that.

**Comment by Idolf** [comment]
I think that removed all of the formatting without removing the actual changed code. I will try not to reintroduce it.

I will add the suggestion from @alev now.

**Comment by Idolf** [comment]
Updated (including the original description of what this PR does)

File: src/platform/nrf_pca10001/lib.rs [hunk 2]
```diff
@@ -46,70 +46,79 @@ impl Firestorm {
...
+            let assert_val: $T = mem::transmute(assert_buf);
+            mem::forget(assert_val);
+
+            // Statically allocate a read-write buffer for the value, write our
+            // initial value into it (without dropping the initial zeros) and
+            // return a reference to it.
+            static mut BUF: [u8; $size] = [0; $size];
             let mut tmp : &mut $T = mem::transmute(&mut BUF);
-            *tmp = $e;
+            ptr::write(tmp as *mut $T, $e);
```
Comment (alevy) [inline src/platform/nrf_pca10001/lib.rs]: Why is this change necessary? I don't mind it, but is it more idiomatic or something?

File: src/platform/nrf_pca10001/lib.rs [hunk 3]
```diff
@@ -46,70 +46,79 @@ impl Firestorm {
...
             tmp
         };
-   }
+    }
 }
 
-pub unsafe fn init<'a>() -> &'a mut Firestorm {
-    use core::mem;
+pub unsafe fn init() -> &'static mut Firestorm {
+    use core::{mem, ptr};
```
Comment (alevy) [inline src/platform/nrf_pca10001/lib.rs]: hrm... since `ptr` isn't used anywhere explicitly in this function (rather only inside of `static_init!`) maybe it should be imported within a scope there? Just would be less confusing IMO

File: src/platform/nrf_pca10001/lib.rs [hunk 4]
```diff
@@ -46,70 +46,79 @@ impl Firestorm {
...
+            ptr::write(tmp as *mut $T, $e);
             tmp
         };
-   }
+    }
 }
 
-pub unsafe fn init<'a>() -> &'a mut Firestorm {
-    use core::mem;
+pub unsafe fn init() -> &'static mut Firestorm {
```
Comment (alevy) [inline src/platform/nrf_pca10001/lib.rs]: why?

File: src/platform/nrf_pca10001/lib.rs [hunk 5]
```diff
@@ -46,70 +46,79 @@ impl Firestorm {
...
     for pin in gpio_pins.iter() {
         pin.set_client(gpio);
     }
 
-    let firestorm : &'static mut Firestorm = mem::transmute(&mut FIRESTORM_BUF);
-    *firestorm = Firestorm {
-        gpio: gpio,
-    };
+    let firestorm: &'static mut Firestorm = mem::transmute(&mut FIRESTORM_BUF);
+    *firestorm = Firestorm { gpio: gpio };
```
Comment (alevy) [inline src/platform/nrf_pca10001/lib.rs]: while we're at it, should probably use `static_init!` here too

File: src/platform/nrf_pca10001/lib.rs (hunk 2 above)
Comment (Idolf) [inline src/platform/nrf_pca10001/lib.rs]: The original code would drop the "value" of tmp, i.e. it would drop the zeros. I don't think any of the values we have used actually had destructor implemented, but if they had, then that would have caused trouble.

At least that is why the change is there. I am currently investigating whether I am right.

**Comment by alevy** [comment]
Overall looks good apart from a few comments. Would also be good to keep lines under 80 characters, but these files in particular aren't great about that anyway, so not worth stressing about in this PR.

File: src/platform/nrf_pca10001/lib.rs (hunk 3 above)
Comment (Idolf) [inline src/platform/nrf_pca10001/lib.rs]: Sure, I will do that (and import mem inside the macro as well, so we do not depend on it from our parent scope).

File: src/platform/nrf_pca10001/lib.rs (hunk 4 above)
Comment (Idolf) [inline src/platform/nrf_pca10001/lib.rs]: As I understand it, the compiler will always infer the most general lifetime that it can -- in this case 'static. So the two are equivalent, but they original way of writing it suggests that something else is meant.

File: src/platform/nrf_pca10001/lib.rs (hunk 5 above)
Comment (Idolf) [inline src/platform/nrf_pca10001/lib.rs]: Will do.

**Comment by alevy** [comment]
For the record, this reduces our memory footprint by ~35%, which is really awesome!

File: src/platform/nrf_pca10001/lib.rs (hunk 4 above)
Comment (alevy) [inline src/platform/nrf_pca10001/lib.rs]: Well, the compiler will infer the lifetime based on how this function is used, not based on the implementation. In this case, it's used from `src/main/main.rs#main` which only uses it in the scope of the main function (although that is practically `'static').

In practice this doesn't really matter though in this case I guess.

File: src/platform/nrf_pca10001/lib.rs (hunk 2 above)
Comment (Idolf) [inline src/platform/nrf_pca10001/lib.rs]: Yup, that does indeed seem to be the case. I think the only reason this every worked was that we depended on undefined behavior in the rust compiler.

Specifically we had we used 0xd4 for the buffer value instead of 0, it seems that we would have caused the destructors to run on our invalid object.

File: src/platform/nrf_pca10001/lib.rs (hunk 4 above)
Comment (Idolf) [inline src/platform/nrf_pca10001/lib.rs]: The compiler will still infer the `'static` lifetime as far as I know, even though it is only used in main. I think I read a blog post about this at some point, I will try finding it.

**Comment by Idolf** [comment]
I messed up my git a bit. It should be back to normal now, and I should have incorporated all of your suggestions.
